internal/vm/hcs/vsmb.go

```diff
@@ -0,0 +1,75 @@
+package hcs
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Microsoft/hcsshim/internal/hcs/resourcepaths"
+	hcsschema "github.com/Microsoft/hcsshim/internal/hcs/schema2"
+	"github.com/Microsoft/hcsshim/internal/requesttype"
+	"github.com/Microsoft/hcsshim/internal/vm"
+)
+
+func vmVSMBOptionsToHCS(options *vm.VSMBOptions) *hcsschema.VirtualSmbShareOptions {
+	return &hcsschema.VirtualSmbShareOptions{
+		ReadOnly:            options.ReadOnly,
+		ShareRead:           options.ShareRead,
+		CacheIo:             options.CacheIo,
+		NoOplocks:           options.NoOplocks,
+		NoDirectmap:         options.NoDirectMap,
+		TakeBackupPrivilege: options.TakeBackupPrivilege,
+		PseudoOplocks:       options.PseudoOplocks,
+		PseudoDirnotify:     options.PseudoDirnotify,
+	}
+}
+
+func (uvm *utilityVM) AddVSMB(ctx context.Context, path string, name string, allowed []string, options *vm.VSMBOptions) error {
+	switch uvm.state {
```

Comment:
Instead of the switch, this can just be implemented separately for `utilityVMBuilder` and `utilityVM`.